src/NHibernate/Loader/Hql/QueryLoader.cs

```diff
@@ -495,6 +497,44 @@ internal IEnumerable GetEnumerable(QueryParameters queryParameters, IEventSource
 			return result;
 		}
 
+		internal IAsyncEnumerable<T> GetAsyncEnumerable<T>(QueryParameters queryParameters, IEventSource session)
+		{
+			CheckQuery(queryParameters);
+
+			return new AsyncEnumerableImpl<T>(
+				InitializeAsync,
+				queryParameters.IsReadOnly(session),
+				_queryTranslator.ReturnTypes,
+				_queryTranslator.GetColumnNames(),
+				queryParameters.RowSelection,
+				_selectNewTransformer ?? queryParameters.ResultTransformer,
+				_queryReturnAliases,
+				session);
+
+			async Task<InitializeResult> InitializeAsync(CancellationToken cancellationToken)
+			{
+				Stopwatch stopWatch = null;
+				if (session.Factory.Statistics.IsStatisticsEnabled)
+				{
+					stopWatch = Stopwatch.StartNew();
+				}
+
+				var command = await PrepareQueryCommandAsync(queryParameters, false, session, cancellationToken).ConfigureAwait(false);
+				var dataReader = await GetResultSetAsync(command, queryParameters, session, null, cancellationToken).ConfigureAwait(false);
+
+				if (stopWatch != null)
+				{
+					stopWatch.Stop();
+					session.Factory.StatisticsImplementor.QueryExecuted("HQL: " + _queryTranslator.QueryString, 0, stopWatch.Elapsed);
+					// NH: Different behavior (H3.2 use QueryLoader in AST parser) we need statistic for orginal query too.
```

Comment:
Don't need this comment here.

Reply:
Now this method is shared for `IEnumerable` and `IAsyncEnumerable` and the comment is removed, but I am not sure whether it is still relevant.